module/zfs/brt.c

```diff
@@ -0,0 +1,1717 @@
+		ASSERT(brt->brt_vdevs != NULL);
+
+		memcpy(vdevs, brt->brt_vdevs,
+		    sizeof (brt_vdev_t) * brt->brt_nvdevs);
+		kmem_free(brt->brt_vdevs,
+		    sizeof (brt_vdev_t) * brt->brt_nvdevs);
+	}
+	for (vdevid = brt->brt_nvdevs; vdevid < nvdevs; vdevid++) {
+		brtvd = &vdevs[vdevid];
+
+		brtvd->bv_vdevid = vdevid;
+		brtvd->bv_initiated = FALSE;
+	}
+
+	BRT_DEBUG("BRT VDEVs expanded from %llu to %llu.", brt->brt_nvdevs,
+	    nvdevs);
+
+	brt->brt_vdevs = vdevs;
+	brt->brt_nvdevs = nvdevs;
+}
+
+static boolean_t
+brt_vdev_lookup(brt_t *brt, brt_vdev_t *brtvd, const brt_entry_t *bre)
+{
+	boolean_t found, unlock;
+	uint64_t idx;
+
+	if (!RW_LOCK_HELD(&brt->brt_lock)) {
+		unlock = TRUE;
+		brt_rlock(brt);
+	} else {
+		unlock = FALSE;
+	}
+
+	idx = bre->bre_offset / brt->brt_rangesize;
+	if (brtvd->bv_refcount != NULL && idx < brtvd->bv_size) {
+		/* VDEV wasn't expanded. */
+		found = brtvd->bv_refcount[idx] > 0;
+	} else {
+		found = FALSE;
+	}
+
+	if (unlock)
+		brt_unlock(brt);
+
+	return (found);
+}
+
+static void
+brt_vdev_addref(brt_t *brt, brt_vdev_t *brtvd, const brt_entry_t *bre,
+    uint64_t dsize)
+{
+	uint64_t idx;
+
+	ASSERT(RW_LOCK_HELD(&brt->brt_lock));
+	ASSERT(brtvd != NULL);
+	ASSERT(brtvd->bv_refcount != NULL);
+
+	idx = bre->bre_offset / brt->brt_rangesize;
+	if (idx >= brtvd->bv_size) {
+		/* VDEV has been expanded. */
+		brt_vdev_realloc(brt, brtvd);
+	}
+
+	ASSERT3U(idx, <, brtvd->bv_size);
+
+	brtvd->bv_totalcount++;
+	brtvd->bv_refcount[idx]++;
+	brtvd->bv_dirty = TRUE;
+	idx = idx / brt->brt_blocksize / 8;
+	BT_SET(brtvd->bv_bitmap, idx);
+
+	brtvd->bv_dsize += dsize;
+	brt->brt_dsize += dsize;
+	if (bre->bre_refcount == 1) {
+		brtvd->bv_drefsize += dsize;
+		brt->brt_drefsize += dsize;
+	}
+#ifdef ZFS_BRT_DEBUG
+	brt_vdev_dump(brt);
+#endif
+}
+
+static void
+brt_vdev_decref(brt_t *brt, brt_vdev_t *brtvd, const brt_entry_t *bre,
+    uint64_t dsize)
+{
+	uint64_t idx;
+
+	ASSERT(RW_WRITE_HELD(&brt->brt_lock));
+	ASSERT(brtvd != NULL);
+	ASSERT(brtvd->bv_refcount != NULL);
+
+	idx = bre->bre_offset / brt->brt_rangesize;
+	ASSERT3U(idx, <, brtvd->bv_size);
+
+	ASSERT(brtvd->bv_totalcount > 0);
+	brtvd->bv_totalcount--;
+	ASSERT(brtvd->bv_refcount[idx] > 0);
+	brtvd->bv_refcount[idx]--;
+	brtvd->bv_dirty = TRUE;
+	idx = idx / brt->brt_blocksize / 8;
+	BT_SET(brtvd->bv_bitmap, idx);
+
+	brtvd->bv_dsize -= dsize;
+	brt->brt_dsize -= dsize;
+	if (bre->bre_refcount == 0) {
+		brtvd->bv_drefsize -= dsize;
+		brt->brt_drefsize -= dsize;
+	}
+#ifdef ZFS_BRT_DEBUG
+	brt_vdev_dump(brt);
+#endif
+}
+
+static void
+brt_vdev_sync(brt_t *brt, brt_vdev_t *brtvd, dmu_tx_t *tx)
+{
+	dmu_buf_t *db;
+	brt_vdev_phys_t *bvphys;
+
+	ASSERT(brtvd->bv_dirty);
+	ASSERT(brtvd->bv_mos_brtvdev != 0);
+	ASSERT(dmu_tx_is_syncing(tx));
+
+	VERIFY0(dmu_bonus_hold(brt->brt_mos, brtvd->bv_mos_brtvdev, FTAG, &db));
+
+	/*
+	 * TODO: Walk through brtvd->bv_bitmap and write only dirty parts.
+	 */
+	dmu_write(brt->brt_mos, brtvd->bv_mos_brtvdev, 0,
```

Comment:
Brian pointed out this might need to be byte swapped so it works properly across endian-ness

Reply:
Should we have a header to go with this? to store `BRT_RANGE_SIZE` and anything else that might change in the future?

Otherwise we are likely to have compatibility issues if we ever want to change the range size

Reply:
You won't be able to change range size once the pool is created. Well, it at least will be hard, as you would need to reread the entire BRT table and recreated refcount table. Currently I store range size in metadata to be able to change the define in the future and don't affect existing pools.